File: src/lib/hazardClassifier.ts
// src/lib/hazardClassifier.ts
// IS 2190:2024 Annex B (Table 6) — Hazard Type Classification
//
// Determines whether a building is Low, Moderate, or High hazard
// based on building height, occupant count, floor area, and
// quantities of flammable gases/liquids.
//
// Rule: Building is classified at the HIGHEST hazard level where
// ANY single criterion places it.

import type { BuildingInput, HazardType } from '@/types';

interface HazardCriteria {
    buildingHeightM: number;
    occupantCount: number;
    totalFloorAreaM2: number;
    flammableGasesLitres: number;
    flammableLiquidsLitres: number;
}

export function determineHazardType(input: BuildingInput): {
    hazardType: HazardType;
    reasons: string[];
    clauseRef: string;
} {
    const reasons: string[] = [];
    const clauseRef = 'IS 2190:2024, Annex B (Table 6)';

    const criteria: HazardCriteria = {
        buildingHeightM: input.buildingHeight,
        occupantCount: input.occupantCount,
        totalFloorAreaM2: input.totalFloorArea,
        flammableGasesLitres: input.flammableGasesLitres ?? 0,
        flammableLiquidsLitres: input.flammableLiquidsLitres ?? 0,
    };

    // HIGH hazard: any single criterion exceeds high threshold
    if (criteria.buildingHeightM > 15)
        reasons.push(`Building height ${criteria.buildingHeightM}m > 15m`);
    if (criteria.occupantCount > 250)
        reasons.push(`Occupant count ${criteria.occupantCount} > 250`);
    if (criteria.totalFloorAreaM2 > 3000)
        reasons.push(`Total floor area ${criteria.totalFloorAreaM2}m² > 3000m²`);
    if (criteria.flammableGasesLitres > 3000)
        reasons.push(`Flammable gases ${criteria.flammableGasesLitres}L > 3000L`);
    if (criteria.flammableLiquidsLitres > 1000)
        reasons.push(`Flammable liquids ${criteria.flammableLiquidsLitres}L > 1000L`);

    if (reasons.length > 0) {
        return { hazardType: 'high', reasons, clauseRef };
    }

    // MODERATE hazard: any criterion in moderate range
    const moderateReasons: string[] = [];
    if (criteria.occupantCount >= 15 && criteria.occupantCount <= 250)
        moderateReasons.push(`Occupant count ${criteria.occupantCount} (15–250)`);
    if (criteria.totalFloorAreaM2 >= 300 && criteria.totalFloorAreaM2 <= 3000)
        moderateReasons.push(`Floor area ${criteria.totalFloorAreaM2}m² (300–3000m²)`);
    if (criteria.flammableGasesLitres >= 500 && criteria.flammableGasesLitres <= 3000)
        moderateReasons.push(`Flammable gases ${criteria.flammableGasesLitres}L (500–3000L)`);
    if (criteria.flammableLiquidsLitres >= 250 && criteria.flammableLiquidsLitres <= 1000)
        moderateReasons.push(`Flammable liquids ${criteria.flammableLiquidsLitres}L (250–1000L)`);

    if (moderateReasons.length > 0) {
        return { hazardType: 'moderate', reasons: moderateReasons, clauseRef };
    }

    // LOW hazard: default
    return {
        hazardType: 'low',
        reasons: ['All criteria within low hazard thresholds'],
        clauseRef,
    };
}

File: backend/hazard_classifier.py
# backend/hazard_classifier.py
# IS 2190:2024 Annex B (Table 6) — Hazard Type Classification
#
# Determines whether a building is Low, Moderate, or High hazard.
# Rule: Building is classified at the HIGHEST hazard level where
# ANY single criterion places it.

from models import BuildingInput, HazardType


def determine_hazard_type(
    inp: BuildingInput,
) -> dict:
    """Return {"hazard_type": ..., "reasons": [...], "clause_ref": ...}."""
    reasons: list[str] = []
    clause_ref = "IS 2190:2024, Annex B (Table 6)"

    height = inp.building_height
    occupants = inp.occupant_count
    area = inp.total_floor_area
    gases = inp.flammable_gases_litres or 0
    liquids = inp.flammable_liquids_litres or 0

    # HIGH hazard: any single criterion exceeds high threshold
    if height > 15:
        reasons.append(f"Building height {height}m > 15m")
    if occupants > 250:
        reasons.append(f"Occupant count {occupants} > 250")
    if area > 3000:
        reasons.append(f"Total floor area {area}m² > 3000m²")
    if gases > 3000:
        reasons.append(f"Flammable gases {gases}L > 3000L")
    if liquids > 1000:
        reasons.append(f"Flammable liquids {liquids}L > 1000L")

    if reasons:
        return {"hazard_type": "high", "reasons": reasons, "clause_ref": clause_ref}

    # MODERATE hazard
    moderate_reasons: list[str] = []
    if 15 <= occupants <= 250:
        moderate_reasons.append(f"Occupant count {occupants} (15–250)")
    if 300 <= area <= 3000:
        moderate_reasons.append(f"Floor area {area}m² (300–3000m²)")
    if 500 <= gases <= 3000:
        moderate_reasons.append(f"Flammable gases {gases}L (500–3000L)")
    if 250 <= liquids <= 1000:
        moderate_reasons.append(f"Flammable liquids {liquids}L (250–1000L)")

    if moderate_reasons:
        return {"hazard_type": "moderate", "reasons": moderate_reasons, "clause_ref": clause_ref}

    # LOW hazard: default
    return {
        "hazard_type": "low",
        "reasons": ["All criteria within low hazard thresholds"],
        "clause_ref": clause_ref,
    }
